test: clarify timewarp grace period griefing attack

Sjors opened the pull request:
Adjust the timewarp test to better illustrate the griefing attack discussed here: https://delvingbitcoin.org/t/timewarp-attack-600-second-grace-period/1326/19 

Changing `MAX_TIMEWARP` to something `> MAX_FUTURE_BLOCK_TIME` in `consensus.h` and `mining_basic.py` will cause the updated test to fail. I'm not proposing such a change here of course. The new test should be useful guidance for pool software developers, for why they really should use `curtime`, or least not ignore `mintime`.

Additionally, if the proposal is changed to make `MAX_TIMEWARP >  MAX_FUTURE_BLOCK_TIME` then this test will break, which could be used to demonstrate there's no such griefing attack anymore.

Originally part of #31600.

Sjors commented:
#31600 landed, ready for review

Prabhat1308 commented on test/functional/mining_basic.py:
```diff
@@ -136,26 +137,45 @@ def test_timewarp(self):
 
-        self.log.info("Create block two hours in the future")
-        self.nodes[0].setmocktime(t + MAX_FUTURE_BLOCK_TIME)
+        self.log.info("Create block MAX_TIMEWARP < t < MAX_FUTURE_BLOCK_TIME in the future")
```
The log mentions `t` to be strictly less than `MAX_FUTURE_BLOCK_TIME`. But in line 165 , the assert statement is for `future<=MAX_FUTURE_BLOCK_TIME`.

Sjors replied:
Changed the log statement to `<=`

Prabhat1308 commented on test/functional/mining_basic.py:
```diff
@@ -166,19 +186,29 @@ def test_timewarp(self):
+        # timestamp to MTP + 1, but doesn't take the new timewarp rule into
+        # account (and ignores mintime).
+        mtp = node.getblock(node.getbestblockhash())["mediantime"] + 1
+        bad_block.nTime = mtp + 1
```
on previous line 199 , `mtp` is set to `node.getblock(node.getbestblockhash())["mediantime"] + 1` . from the rpc commands , shouldnt `mtp` be just `node.getblock(node.getbestblockhash())["mediantime"]` ? and then `nTime` is correct with `mtp + 1`. Have checked it locally and the test passes.

Prabhat1308 commented on test/functional/mining_basic.py:
```diff
@@ -136,26 +137,45 @@ def test_timewarp(self):
+        # The template will have an adjusted timestamp.
         tmpl = node.getblocktemplate(NORMAL_GBT_REQUEST_PARAMS)
-        assert_greater_than_or_equal(tmpl['curtime'], t + MAX_FUTURE_BLOCK_TIME - MAX_TIMEWARP)
+        assert_equal(tmpl['curtime'], future - MAX_TIMEWARP)
```
this is introducing a more stricter equality. Is it because we now have the liberty to choose `future` as we wish whereas previously it was `t + MAX_FUTURE_BLOCK_TIME` which not necessarily be equal to template's curtime ?

Sjors replied:
> more stricter equality.

Stricter than what? I don't fully remember what the original test was trying to do here. But we know the exact value to expect.

`curtime` is equal to `nTime` of the template, which is controlled by `UpdateTime()` in `node/miner.cpp`. It starts with the actual current time, but then increases if needed to account for the MTP+1 rule and timewarp - and no more than that.

Prabhat1308 reviewed (commented):
ACK [`6065b78`](https://github.com/bitcoin/bitcoin/pull/31725/commits/6065b78e22e47f053227f68afb8e3afa7bf0ab95)

Sjors committed:
test: clarify timewarp griefing attack

On testnet4 with the timewarp mitigation active, when pool software
ignores the curtime and mintime fields provided by the getblocktemplate
RPC or by createNewBlock() in the Mining interface, they are vulnerable
to a griefing attack.

The test is expanded to illustrate this.

Prabhat1308 commented:
re-ACK [`37d2246`](https://github.com/bitcoin/bitcoin/pull/31725/commits/37d22461aa8f62d7fee9b65c8580b7f7a25f3382)

darosior reviewed (commented):
This is adding a test that demonstrates how Bitcoin Core would reject a block from a broken mining pool software that:
- ignores the block chain and set the block timestamp using wall clock time;
- ignores the `mintime` value provided by Bitcoin Core's block template;
- ignores the `curtime` value provided by Bitcoin Core's block template.

There is many, many ways a mining pool software that disregards consensus rules and values provided by Bitcoin Core's block template may create an invalid block. I don't think it's useful to exercise one such convoluted scenario in particular in Bitcoin Core's functional test suite. For this reason i am by default a (weakly held) Concept NACK on this change.

Weakly held because it's simply a test change and i'm happy to be convinced otherwise.

Sjors commented:
[BIP54](https://github.com/bitcoin/bips/blob/master/bip-0054.md) has been updated to use a 2 hour grace period rather than the `MAX_TIMEWARP = 600;` which our code still uses.

If you change `MAX_TIMEWARP` to `7200` there's no way to make the test in this PR pass. I'll leave that as an exercise to the reader.

Note that the list of three bullet items boils down to _a single mistake_: using wall time for `nTime`. And as documented in the Delving post, that's not a hypothetical mistake. It's no longer a problem because BIP54 uses the longer grace period.
